New demo scene using Unity Terrain

williamtsoi opened the pull request:
Hi Reese,
Thanks for making this open source repo! 

I'm working on an ECS project using Unity Terrain, and wanted to create a quick demo with this to get pathfinding working. but I seem to be running into some performance issues.

Here is the original performance demo, spawned 2k ish agents and I'm getting 155fps (they all pathed away by the time i took the screenshot but they did spawn!)
![2020-07-03 (6)](https://user-images.githubusercontent.com/2007359/86495830-d7a41e80-bd72-11ea-8f54-5ef54b2fc581.png)

And here is the terrain demo scene, where it is struggling at 25fps for 200ish entities.
![2020-07-03 (2)](https://user-images.githubusercontent.com/2007359/86495784-b04d5180-bd72-11ea-89ae-e5b721955c61.png)

Here's the profiler, but I'm not sure where to look or what this really means though.
![2020-07-03 (5)](https://user-images.githubusercontent.com/2007359/86495852-e1c61d00-bd72-11ea-8bfa-7a73558d7704.png)

Any chance you might be able to have a look or point me in the right direction for this?

Thanks in advance :)

williamtsoi commented:
Thanks for your pointer towards the invalid destination issue. Looks like that was definitely at least part of the issue.

I've added a check using a raycast for the randomly generated destination to make sure it lands onto the terrain collider before adding it as a `NavNeedsDestination` component and that seems to have helped. 
Also added a check for non successful `NavPlanning`s and to remove the `NavPlanning` tag so it will look for a new one, is this what you mean for #35 ?

Also do let me know if you'd like me to comment/commit in a different style etc, first time collaborating on a project, and not sure if there is a standard to how how things are done.

williamtsoi commented on Assets/Scripts/Nav/Destination/NavTerrainDestinationSystem.cs:
```diff
@@ -20,17 +25,21 @@ protected override void OnCreate()
 
         protected override void OnUpdate()
         {
+            var collisionWorld = buildPhysicsWorld.PhysicsWorld.CollisionWorld;
```
Not sure when to us e CollisionWorld vs PhysicsWorld, but the [docs examples](https://docs.unity3d.com/Packages/com.unity.physics@0.4/manual/collision_queries.html) are in CollisionWorld so thats what I've used here, but noticed the rest of your project was using PhysicsWorld, so I can change this to be consistent if you would prefer.

reeseschultz replied:
Functionally I'm not sure what the difference is between the two, and have been wondering about that as well. I may have to dig into Unity's code to find out. One may be more performant than the other, but I'm not sure which. For consistency with this project, you can just use the physics world, but if it turns out using the collision world is objectively better somehow, then we can always refactor everything in one fell swoop later.

williamtsoi replied:
Agreed - updated.

reeseschultz commented:
For #35, you are correct that the `NavPlanning` tag ought to be removed upon failure. `NavNeedsDestination` may need removal too, depending on how you want to handle things conditionally to get a new destination. What I do intend to do with #35 later is passthrough different failure statuses, so the user knows how the planning went in detail. For example, a user who receives an `OutOfNodes` error may want to leverage a procedure that adds nodes in that case. `WrongMagic` might imply that the nav mesh needs re-baking, possibly. Multiple flags can be set, complicating matters, so they may just have to be passed through as bit flags in a new component. I think the change you already made should suffice for now, and will partially account for #35. For now we can just assume if `NavPlanning` component is removed, but `NavNeedsDestination` exists and there is no `NavLerping`, something must have went wrong.

If you want to handle that more semantically, feel free to add code as you see fit. Your reasoning looks great to me so far. In terms of style, by the way, don't worry much about that either. I can make edits when the PR is done-ish, and then we can squash everything upon merge. We can also figure out how you might want credit for your changes. Also, thank you! I really appreciate having help on this project.

williamtsoi commented:
Thanks :) cleaned up a bit based on your comments, also renamed it to `GetPointOnSurfaceLayer` to make it more explicit. Looks like Extension methods only work with `this` so kept it as that, but moved it into `NavUtil`. Not done much else other than that yet.

Its its now getting similar performance to the performance demo, which I'm happy about. I think next step for me is to see about getting the agents aware of the terrain height so they don't sync into the ground (or fly? not sure about that one on the right). I'm thinking adding a collider to the agents and use that to set the `y` to see if that will still run smoothly.
![2020-07-07_LI_Moment_Moment](https://user-images.githubusercontent.com/2007359/86851029-266ef280-c0aa-11ea-87f2-44e0d0f07fbe.jpg)

Also currently I am actually manually creating the `RenderBounds` for the terrain in order for the spawner to pick random locations, ideally I need to figure out how to actually get a valid render bounds for a terrain, but I think this is lower priority since the spawning isn't directly related to the pathing, I guess another method can be used for picking locations if it comes down to that.

williamtsoi commented:
Hi Reese,

Totally agree that Raycasting every frame shouldn't be the way to go. Was just surprised that it was actually running smoothly even with 1000k+ agents.

I was looking in the `NavInterpolationSystem` but it looks like the the obstacle avoidance is only related to jumping between surfaces so that doesn't seem to be able to either help me or cause me issues regarding this.

It seems that the cause is related to how the paths are generated between points on the Navmesh.
The path is a straight lines between points on the mesh, and this could mean that the straight line actually cuts through the terrain since the height between the two points can vary by a lot. The `NavInterpolationSystem` then uses the way point to move forward therefor also following that same line and cutting through terrain.
So I don't think there is an actual bug here per say, just that it cannot handle uneven terrain as is.
Because of this, I also don't think casting only at the start of path will fix it either, the problem will only manifest itself mid path and actually would be fine at start and end of paths.

Hope that makes sense.
![image](https://user-images.githubusercontent.com/2007359/87464755-8f56ed00-c60b-11ea-9278-3e799a15f49d.png)
![image](https://user-images.githubusercontent.com/2007359/87465466-9f230100-c60c-11ea-81df-3dafd3a63bb3.png)

What do you think of using a Collider for this instead of the Raycasting approach, or prehaps a throttled Raycast? 
I think the standard Unity agent uses a simplified Raycast as well for following the terrain, guessing by looking at this method name: `public bool SamplePathPosition(int areaMask, float maxDistance, out NavMeshHit hit);` [unity-decompiled/../NavMeshAgent.cs](https://github.com/jamesjlinden/unity-decompiled/blob/master/UnityEngine/UnityEngine/NavMeshAgent.cs)

I think maybe an optional tag that could be added a NavAgent entity which would allow it to use an existing collider on the entity if it already contains one, or Raycast as a fall back if it doesn't? Depending on the project, NavAgents might or might not already have an existing collider.

Let me know what you think, Cheers

reeseschultz commented:
You raise good points, @williamtsoi. Thanks for bearing with me.

> I think maybe an optional tag that could be added a NavAgent entity which would allow it to use an existing collider on the entity if it already contains one, or Raycast as a fall back if it doesn't?

I like that a lot. I think adding a `NavTerrainCapable` component tag to `NavAgentStatus.cs` would do the trick. In the `NavGroundedAgentSystem`, it can query only agents that have said tag. You can define the tag as follows:

```csharp
/// <summary>Exists if the agent needs to traverse complex terrain.</summary>
public struct NavTerrainCapable : IComponentData { }
```

When spawning agents for the terrain demo, just add that tag to them and you should be good to go.

Beyond that, I don't think any further work would be required for this PR, unless there's more you feel needs doing. I don't think the optimizations are necessary for us to merge this PR if we leverage that component tag. Having it alleviates my concerns since then I don't have to worry about adverse effects on the other demos/use cases.

If interested, you could submit a subsequent PR later to potentially optimize the grounding for terrain. I like the idea of separate jobs for `NavTerrainCapable` agents that have `PhysicsCollider` components and those that do not, but the work involved may count as a feature in and of itself. And speculative throttled raycasting may or may not be yet another PR. We would want to confirm that the work really had a positive performance impact anyway. You would be under no obligation to submit any further PRs, but you're more than welcome to do so. I think the work you're doing here is fantastic.

williamtsoi commented:
>I think maybe an optional tag that could be added a NavAgent entity which would allow it to use an existing collider on the entity if it already contains one, or Raycast as a fall back if it doesn't?

So after some investigation, looks like using a collider wont work how I imagined it could, the collision [Trigger Event](https://docs.unity3d.com/Packages/com.unity.physics@0.4/api/Unity.Physics.TriggerEvent.html), doesn't actually return the collision point, and the doing a collider cast would be more expensive than the ray cast, so I've removed that implementation and returned it to only using a Raycast again. I've added the optional tag for the `NavGroundedAgentSystem` and updated the spawner so it's isolated to the the terrain demo.

So the last thing I've I've been trying to add was to make the agent align the the surface normal - think cars, tanks etc.
I got close over the last few days, but I can't work out how to make it work correctly.
The current solution rotates it along the Z axis correctly, but not the X axis, so the agent is rotated sideways correctly, but not looking down the hill like it should.
The green line is the surface normal, the drawn blue line is where it should be looking, but it is instead facing forward on the blue+white line.
![image](https://user-images.githubusercontent.com/2007359/87885960-4343e800-ca11-11ea-9563-69f5f938e127.png)

I've given it a good attempt but I don't think I'll be able to work this one out at the moment, would you rather I pulled this change out so we can complete this PR, or would you be able to maybe have a look or point me in the right direction?

reeseschultz commented:
@williamtsoi I got you fam.

Getting the rotation right is a multi-step process. In other words, a pain in the ass. Here's one way to do it:

1. In the `NavAgent`, add this member:

```csharp
/// <summary>Normal of point below the agent.</summary>
public float3 SurfacePointNormal;
```

2. In the `NavAgentGroundedSystem`, now all you need is this:

```csharp
if (physicsWorld.CastRay(rayInput, out RaycastHit hit)) agent.SurfacePointNormal = hit.SurfaceNormal;
```

(But you're welcome to leave your debug code in place, by the way, I think it's useful. )

3. We'll keep all our interpolation in the `NavInterpolationSystem` not just because of the [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle), but to get the rotation right, so the end of the `NavWalkJob` should be replaced with this:

```csharp
translation.Value = Vector3.MoveTowards(translation.Value, localWaypoint, agent.TranslationSpeed * deltaSeconds);

var lookAt = NavUtil.MultiplyPoint3x4( // To world (from local in terms of destination surface).
    localToWorldFromEntity[agent.DestinationSurface].Value,
    localWaypoint
);

lookAt = NavUtil.MultiplyPoint3x4( // To local (in terms of agent's current surface).
    math.inverse(localToWorldFromEntity[surface.Value].Value),
    lookAt
);

lookAt.y = translation.Value.y;

var lookRotation = quaternion.LookRotationSafe(lookAt - translation.Value, math.up());

if (((Vector3)agent.SurfacePointNormal).magnitude > 0.1f)
    lookRotation = Quaternion.FromToRotation(localToWorldFromEntity[entity].Up, agent.SurfacePointNormal) * lookRotation;

rotation.Value = math.slerp(rotation.Value, lookRotation, deltaSeconds / 0.3f);
```
We're using both `quaternion` and `Quaternion` with a capital Q since for some reason there isn't feature parity between the two. To get around that I was going to use `Unity.Physics.Math.FromToRotation` for calculating the normal rotation, but it doesn't work as I expect, so we can just live with what amounts to an annoying style discrepancy for the sake of getting this working for now.

You'll note two magic numbers. The one used in the `SurfacePointNormal` magnitude check (`0.1f`) is probably fine. Although, you may want to add a rotation speed member called `RotationSpeed` to the `NavAgent` for the `slerp` operation, replacing the `0.3f` with it.

Other things I want to bring up:
* I think we should change the name of `NavGroundedAgentSystem` to `NavGroundingSystem` for consistency with the other systems (using present as opposed to past tense).
* I updated a bunch of stuff in the main branch, but I can take care of merging if you're not comfortable with that.

People will dig this feature/demo. Honestly they'll probably find this more useful than the transformation hierarchy stuff that I obsess over. Thanks for sticking with this.

williamtsoi commented:
Thanks @reeseschultz , that helps a lot!

Didn't think of using the standard Quaternion for help.
Just had to tweak this line slightly to make it rotate fully to the normal `lookRotation = Quaternion.FromToRotation(math.up(), agent.SurfacePointNormal) * lookRotation;`.

Just wanted to do some final tests with the scene before signing this off so I added more unpathable areas to the terrain. 
Unfortunately I've noticed that the agents seems to be getting stuck along some of the edges and don't seem to get close enough to the `localWaypoint` for it to move onto the next waypoint. -at least that's what I think is happening from the initial investigation. I'll be digging into this some more. 

Hoping to get this hopefully-final issue resolved soon :)
Cheers

reeseschultz commented on Assets/Scripts/Nav/NavTerrainDemo/DebugMode.cs:
```diff
@@ -37,13 +37,13 @@ private void OnEnable()
         {
-            groundedAgentSystem.IsDebugging = IsDebugging = isDebugging;
-            groundedAgentSystem.DrawUnitVectors = drawUnitVectors;
+            navGroundingSystem.IsDebugging = IsDebugging = isDebugging;
```
That's a lot of flags for debugging. Might it be possible to reduce them, preferably down to one?

williamtsoi replied:
Good point, updated and kept the toggles separate in the inspector, but clasped down to 1 flag per system that is being debugged.

reeseschultz commented on Assets/Scripts/Nav/Spawner/NavTerrainSpawner.cs:
```diff
@@ -79,6 +79,7 @@ void Enqueue()
                         JumpDegrees = 45,
                         JumpGravity = 200,
                         TranslationSpeed = 20,
+                        RotationSpeed = 0.3f,
```
Would you mind updating the `RotationSpeed` to `0.3f` in all the other nav-related spawners as well?

reeseschultz commented on Packages/com.reese.nav/Agent/NavGroundingSystem.cs:
```diff
@@ -43,26 +43,23 @@ protected override void OnUpdate()
                    }
                })
-               .WithName("Set_grounded_height_and_rotation")
+               .WithName("GroundingJob")
```
That looks much better!

reeseschultz commented on Packages/com.reese.nav/Agent/NavInterpolationSystem.cs:
```diff
@@ -102,9 +104,13 @@ protected override void OnUpdate()
-                    translation.Value = Vector3.MoveTowards(translation.Value, localWaypoint, agent.TranslationSpeed * deltaSeconds);
+                    var lookRotation = quaternion.LookRotationSafe(lookAt - translation.Value, math.up());
+
+                    if (math.length(agent.SurfacePointNormal) > 0.01f)
```
Good! I didn't know about `math.length`.

reeseschultz commented on Packages/com.reese.nav/Agent/NavPlanSystem.cs:
```diff
@@ -1,168 +1,172 @@
+                        NavConstants.ITERATION_MAX,
+                        out int iterationsPerformed
+                    );
+
```
Looks like there's an unintended extra newline here. Also, I'm curious, were all the line endings in this file changed?

williamtsoi replied:
I'm not sure about the line ending, looks like it is currently CRLF just like the other files, not sure why it is flagging it whole file as modified here.

reeseschultz commented on Packages/com.reese.nav/NavUtil.cs:
```diff
@@ -1,101 +1,102 @@
+            return position;
+        }
+
+        /// <summary>Extention method for PhysicsWorld to check for a valid
```
Please replace "Extention" with "Extension."

reeseschultz commented on Packages/com.reese.nav/NavUtil.cs:
```diff
@@ -1,101 +1,102 @@
+        }
+
+        /// <summary>Extention method for PhysicsWorld to check for a valid
+        /// position using raycast onto the surface layer based on the position
```
Please replace "using raycast" with "by raycasting."

reeseschultz commented on Packages/com.reese.nav/NavUtil.cs:
```diff
@@ -1,101 +1,102 @@
+            var rayInput = new RaycastInput()
+            {
+                Start = position + localToWorld.Up * NavConstants.OBSTACLE_RAYCAST_DISTANCE_MAX,
+                End = position + -localToWorld.Up * NavConstants.OBSTACLE_RAYCAST_DISTANCE_MAX,
```
Please replace `position + -localToWorld.Up` with `position - localToWorld.Up`.

reeseschultz commented on Packages/com.reese.nav/NavUtil.cs:
```diff
@@ -1,101 +1,102 @@
+
+        /// <summary>Extention method for PhysicsWorld to check for a valid
+        /// position using raycast onto the surface layer based on the position
+        /// passed in. bool returned based on if the raycast was successful and 
```
Please replace "was successful" with "is successful."

reeseschultz commented on Packages/com.reese.nav/NavUtil.cs:
```diff
@@ -1,101 +1,102 @@
+        /// <summary>Extention method for PhysicsWorld to check for a valid
+        /// position using raycast onto the surface layer based on the position
+        /// passed in. bool returned based on if the raycast was successful and 
+        /// Valid position returned in the Out parameter same as the standard 
```
Please replace this and the next line (42 and 43) with `/// a position returned via out.`

reeseschultz commented on Packages/com.reese.nav/NavUtil.cs:
```diff
@@ -1,101 +1,102 @@
+                {
+                    BelongsTo = NavUtil.ToBitMask(NavConstants.COLLIDER_LAYER),
+                    CollidesWith = NavUtil.ToBitMask(NavConstants.SURFACE_LAYER),
+                    GroupIndex = 0
```
You can remove this line and the comma on the line above (53) since the `GroupIndex` default is equivalent.

williamtsoi commented on Assets/Scripts/Nav/NavTerrainDemo/DebugMode.cs:
```diff
@@ -38,26 +40,49 @@ private void OnEnable()
+                    {
+                        var node = pathBuffer[i];
+
+                        Debug.DrawLine(node.Value + drawPathOffset, pathBuffer[i + 1].Value + drawPathOffset, Color.red);
```
Didn't realise I could use Debug.DrawLine from a System as well now, so moved this into a SystemBase Entities loop.

williamtsoi commented on Packages/com.reese.nav/Agent/NavGroundingSystem.cs:
```diff
@@ -55,12 +53,12 @@ protected override void OnUpdate()
 
                        var currentPosition = translation.Value;
-                       currentPosition.y = hit.Position.y + agent.Offset.y;
+                       currentPosition.y = hit.Position.y;
```
Should not be setting the Agent's physical position by the offset. I think I misunderstood what the offset was for.

williamtsoi commented on Packages/com.reese.nav/Agent/NavInterpolationSystem.cs:
```diff
@@ -58,7 +58,7 @@ protected override void OnUpdate()
 
                     if (
-                        NavUtil.ApproxEquals(translation.Value, localWaypoint, 1) &&
+                        NavUtil.ApproxEquals(translation.Value + agent.Offset, localWaypoint, 1) &&
```
Added offset when checking for distance to current waypoint, same with the checks in the rest of the file.

Looks like this was 1 of the issue which was causing the agent to stop mid path as the vertical offset was too great from the path's y position.

williamtsoi commented on Packages/com.reese.nav/NavConstants.cs:
```diff
@@ -42,7 +42,7 @@ public static class NavConstants
         /// increased if an OutOfNodes error arises while finding a path.
         /// </summary>
-        public const int NAV_MESH_QUERY_NODE_MAX = 1000;
+        public const int NAV_MESH_QUERY_NODE_MAX = 5000;
```
Looks like this was was the cause of the other issue I was having.
When the nav mesh was made more complex due to the terrain having more unpathable areas, it was running out of nodes to find a valid path.

williamtsoi commented:
Hi @reeseschultz , sorry its been a while since I was able to have a look at this. I've finally been able to resolve that final issue I was having with agents getting stuck on way points. 
Think this might be ready for the final review!

reeseschultz commented:
![](https://media.giphy.com/media/8vpeyWA3OWOhG/giphy.gif)

Thank you so much for your effort @williamtsoi, you did a great job. I'm merging this into a temporary feature branch to make some minor adjustments, then it will be merged into the main branch afterward.
